src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPDocumentMappingProvider.cs

```diff
@@ -58,5 +70,148 @@ public async override Task<RazorMapToDocumentRangeResponse> MapToDocumentRangeAs
         }
+
+        public async override Task<WorkspaceEdit> RemapWorkspaceEditAsync(WorkspaceEdit workspaceEdit, CancellationToken cancellationToken)
+        {
+            if (workspaceEdit == null)
+            {
+                return workspaceEdit;
+            }
+            else if (workspaceEdit.DocumentChanges != null)
+            {
+                // The LSP spec says, we should prefer `DocumentChanges` property over `Changes` if available.
+                var remappedEdits = await RemapVersionedDocumentEditsAsync(workspaceEdit.DocumentChanges, cancellationToken).ConfigureAwait(false);
+                return new WorkspaceEdit()
+                {
+                    DocumentChanges = remappedEdits
+                };
+            }
+            else
+            {
+                var remappedEdits = await RemapDocumentEditsAsync(workspaceEdit.Changes, cancellationToken).ConfigureAwait(false);
+                return new WorkspaceEdit()
+                {
+                    Changes = remappedEdits
+                };
+            }
+        }
+
+        private async Task<TextDocumentEdit[]> RemapVersionedDocumentEditsAsync(TextDocumentEdit[] documentEdits, CancellationToken cancellationToken)
+        {
+            var remappedDocumentEdits = new List<TextDocumentEdit>();
+            foreach (var entry in documentEdits)
+            {
+                var uri = entry.TextDocument.Uri;
+                if (!CanRemap(uri))
+                {
+                    // This location doesn't point to a background razor file. No need to remap.
+                    remappedDocumentEdits.Add(entry);
+
+                    continue;
+                }
+
+                var edits = entry.Edits;
+                var (documentSnapshot, remappedEdits) = await RemapTextEditsAsync(uri, edits, cancellationToken).ConfigureAwait(false);
+                if (documentSnapshot == null)
+                {
+                    // Couldn't find the document. Ignore this edit.
+                    continue;
+                }
+
+                remappedDocumentEdits.Add(new TextDocumentEdit()
+                {
+                    TextDocument = new VersionedTextDocumentIdentifier()
+                    {
+                        Uri = documentSnapshot.Uri,
+                        Version = documentSnapshot.Version
+                    },
+                    Edits = remappedEdits
+                });
+            }
+
+            return remappedDocumentEdits.ToArray();
+        }
+
+        private async Task<Dictionary<string, TextEdit[]>> RemapDocumentEditsAsync(Dictionary<string, TextEdit[]> changes, CancellationToken cancellationToken)
+        {
+            var remappedChanges = new Dictionary<string, TextEdit[]>();
+            foreach (var entry in changes)
+            {
+                var uri = new Uri(entry.Key);
+                var edits = entry.Value;
+
+                if (!CanRemap(uri))
```

Comment:
Perhaps some abstraction here which allows for `CanRemap` to take into account whether we support RazorCSharp, RazorHTML, or both. For ex. for Find all Definitions we'd only want this check to verify if it's a RazorCSharp file.

Reply:
Disallowing RazorHTML from mapping is just a shortcut we are taking because we happen to map the virtual html and the Razor locations 1:1. But in my mind that is just an implementation detail and could change. So I made it so that both RazorC# and RazorHTML documents are remapped regardless.